File: src/test/java/com/example/forum/controller/DataRestTest.java
package com.example.forum.controller;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@DisplayName("Data Rest - API 테스트")
@Transactional
@AutoConfigureMockMvc
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
public class DataRestTest {
    private final MockMvc mvc;

    public DataRestTest(@Autowired MockMvc mvc) {
        this.mvc = mvc;
    }

    @DisplayName("게시글 리스트 조회")
    @Test
    @WithMockUser
    void givenNothing_whenRequestingArticles_thenReturnsArticlesJSONResponse() throws Exception {
        // Given

        // When & Then
        mvc.perform(get("/api/articles"))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.valueOf("application/hal+json")))
                .andDo(print());
    }

    @DisplayName("게시글 단건 조회")
    @Test
    @WithMockUser
    void givenNothing_whenRequestingArticles_thenReturnsArticlesOnlyOne() throws Exception {
        // Given

        // When & Then
        mvc.perform(get("/api/articles/3"))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.valueOf("application/hal+json")));
    }

    @DisplayName("게시글 -> 댓글 리스트 조회")
    @Test
    @WithMockUser
    void givenNothing_whenRequestingArticleCommentsFromArticle_thenReturnsArticleCommentsJsonResponse() throws Exception {
        // Given

        // When & Then
        mvc.perform(get("/api/articles/3/articleComments"))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.valueOf("application/hal+json")));
    }


    @DisplayName("댓글 리스트 조회")
    @Test
    @WithMockUser
    void givenNothing_whenRequestingArticleComments_thenReturnsArticleCommentsJsonResponse() throws Exception {
        // Given

        // When & Then
        mvc.perform(get("/api/articles/5/articleComments"))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.valueOf("application/hal+json")));
    }


    @DisplayName("댓글 단건 조회")
    @Test
    @WithMockUser
    void givenNothing_whenRequestingArticleComment_thenReturnsArticleCommentJsonResponse() throws Exception {
        // Given

        // When & Then
        mvc.perform(get("/api/articleComments/1"))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.valueOf("application/hal+json")));
    }
}

File: migration-to-kotlin/src/test/kotlin/com/example/practice/controller/DataRestTest.kt
package com.example.practice.controller

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import org.springframework.transaction.annotation.Transactional

@DisplayName("Data Rest - API Test")
@Transactional
@WithMockUser
@AutoConfigureMockMvc
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
class DataRestTest @Autowired constructor(private val mvc : MockMvc) {
    @DisplayName("게시글 리스트 조회")
    @Test
    internal fun givenNothing_whenRequestingArticles_thenReturnsArticlesJSONResponse() {
        mvc.perform(get("/api/articles"))
            .andExpect(status().isOk)
            .andExpect(content().contentType(MediaType.valueOf("application/hal+json")))
            .andDo(print())
    }

    @DisplayName("게시글 단건 조회")
    @Test
    internal fun givenNothing_whenRequestingArticles_thenReturnsArticlesOnlyOne() {
        mvc.perform(get("/api/articles/1"))
            .andExpect(status().is4xxClientError)
            //.andExpect(content().contentType(MediaType.valueOf("application/hal+json")))
    }

    @DisplayName("게시글 -> 댓글 리스트 조회")
    @Test
    internal fun givenNothing_whenRequestingArticleCommentsFromArticle_thenReturnsArticleCommentsJsonResponse() {
        mvc.perform(get("/api/articles/3/articleComments"))
            .andExpect(status().is4xxClientError)
    }

    @DisplayName("댓글 리스트 조회")
    @Test
    internal fun givenNothing_whenRequestingArticleComments_thenReturnsArticleCommentsJsonResonse() {
        mvc.perform(get("/api/articleComments"))
            .andExpect(status().is4xxClientError)
    }

    @DisplayName("댓글 단건 조회")
    @Test
    internal fun givenNothing_whenRequestingArticleComment_thenReturnsArticleCommentJsonResponse() {
        mvc.perform(get("/api/articleComments/1"))
            .andExpect(status().is4xxClientError)
    }
}